File: facebook/src/com/facebook/BoltsMeasurementEventListener.java
package com.facebook;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Bundle;
import android.support.v4.content.LocalBroadcastManager;

public class BoltsMeasurementEventListener extends BroadcastReceiver {
    private static BoltsMeasurementEventListener _instance;

    private final static String MEASUREMENT_EVENT_NOTIFICATION_NAME = "com.parse.bolts.measurement_event";
    private final static String MEASUREMENT_EVENT_NAME_KEY = "event_name";
    private final static String MEASUREMENT_EVENT_ARGS_KEY = "event_args";
    private final static String BOLTS_MEASUREMENT_EVENT_PREFIX = "bf_";

    private Context applicationContext;

    private BoltsMeasurementEventListener(Context context) {
        applicationContext = context.getApplicationContext();
    }

    private void open() {
      LocalBroadcastManager broadcastManager = LocalBroadcastManager.getInstance(applicationContext);
      broadcastManager.registerReceiver(this, new IntentFilter(MEASUREMENT_EVENT_NOTIFICATION_NAME));
    }

    private void close() {
      LocalBroadcastManager broadcastManager = LocalBroadcastManager.getInstance(applicationContext);
      broadcastManager.unregisterReceiver(this);
    }

    static BoltsMeasurementEventListener getInstance(Context context) {
        if (_instance != null) {
            return _instance;
        }
        _instance = new BoltsMeasurementEventListener(context);
        _instance.open();
        return _instance;
    }

    protected void finalize() throws Throwable {
        try {
            close();
        } finally {
            super.finalize();
        }
    }

    @Override
    public void onReceive(Context context, Intent intent) {
        AppEventsLogger appEventsLogger = AppEventsLogger.newLogger(context);
        String eventName = BOLTS_MEASUREMENT_EVENT_PREFIX + intent.getStringExtra(MEASUREMENT_EVENT_NAME_KEY);
        Bundle eventArgs = intent.getBundleExtra(MEASUREMENT_EVENT_ARGS_KEY);
        Bundle logData = new Bundle();
        for(String key : eventArgs.keySet()) {
           String safeKey = key.replaceAll("[^0-9a-zA-Z _-]", "-").replaceAll("^[ -]*", "").replaceAll("[ -]*$", "");
           logData.putString(safeKey, (String)eventArgs.get(key));
        }
        appEventsLogger.logEvent(eventName, logData);
    }
}

File: facebook-core/src/main/java/com/facebook/internal/BoltsMeasurementEventListener.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.internal

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Bundle
import androidx.annotation.VisibleForTesting
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import com.facebook.appevents.InternalAppEventsLogger
import com.facebook.internal.instrument.crashshield.AutoHandleExceptions

/**
 * com.facebook.internal is solely for the use of other packages within the Facebook SDK for
 * Android. Use of any of the classes in this package is unsupported, and they may be modified or
 * removed without warning at any time.
 */
@AutoHandleExceptions
class BoltsMeasurementEventListener private constructor(context: Context) : BroadcastReceiver() {
  private val applicationContext: Context = context.applicationContext
  companion object {
    private var singleton: BoltsMeasurementEventListener? = null

    @VisibleForTesting
    internal val MEASUREMENT_EVENT_NOTIFICATION_NAME = "com.parse.bolts.measurement_event"
    private const val MEASUREMENT_EVENT_NAME_KEY = "event_name"
    private const val MEASUREMENT_EVENT_ARGS_KEY = "event_args"
    private const val BOLTS_MEASUREMENT_EVENT_PREFIX = "bf_"

    @JvmStatic
    fun getInstance(context: Context): BoltsMeasurementEventListener? {
      if (singleton != null) {
        return singleton
      }
      val listener = BoltsMeasurementEventListener(context)
      listener.open()
      singleton = listener
      return singleton
    }
  }

  private fun open() {
    val broadcastManager = LocalBroadcastManager.getInstance(applicationContext)
    broadcastManager.registerReceiver(this, IntentFilter(MEASUREMENT_EVENT_NOTIFICATION_NAME))
  }

  private fun close() {
    val broadcastManager = LocalBroadcastManager.getInstance(applicationContext)
    broadcastManager.unregisterReceiver(this)
  }

  @Throws(Throwable::class)
  fun finalize() {
    close()
  }

  override fun onReceive(context: Context?, intent: Intent?) {
    val logger = InternalAppEventsLogger(context)
    val eventName =
        BOLTS_MEASUREMENT_EVENT_PREFIX + intent?.getStringExtra(MEASUREMENT_EVENT_NAME_KEY)
    val eventArgs = intent?.getBundleExtra(MEASUREMENT_EVENT_ARGS_KEY)
    val logData = Bundle()
    val keySet = eventArgs?.keySet()
    if (keySet != null) {
      for (key in keySet) {
        val safeKey =
            key.replace("[^0-9a-zA-Z _-]".toRegex(), "-")
                .replace("^[ -]*".toRegex(), "")
                .replace("[ -]*$".toRegex(), "")
        logData.putString(safeKey, eventArgs.get(key) as String?)
      }
    }
    logger.logEvent(eventName, logData)
  }
}
